ts/private/ts_project.bzl

```diff
@@ -215,81 +215,113 @@ This is an error because Bazel does not run actions unless their outputs are nee
 
     output_types = typings_outs + typing_maps_outs + typings_srcs
 
+    # What tsc will be emitting
+    use_tsc_for_js = len(js_outs) > 0
+    use_tsc_for_dts = len(typings_outs) > 0
```

Comment:
Due to the reason I mentioned in my comment above, this will be incorrect (non-empty) when using a custom `declaration_transpiler`.

Reply:
I think this is the last thing left. I added this patch on top of the latest version of this branch and everything seems to be working.
```diff
diff --git a/ts/defs.bzl b/ts/defs.bzl
index 63f885d..875f768 100644
--- a/ts/defs.bzl
+++ b/ts/defs.bzl
@@ -481,6 +481,7 @@ def ts_project(
         tsc = tsc,
         tsc_worker = tsc_worker,
         transpile = -1 if not transpiler else int(transpiler == "tsc"),
+        declaration_transpile = -1 if not declaration_transpiler else int(declaration_transpiler == "tsc"),
         supports_workers = supports_workers,
         is_typescript_5_or_greater = is_typescript_5_or_greater,
         validate = validate,
diff --git a/ts/private/ts_lib.bzl b/ts/private/ts_lib.bzl
index 99cab36..1d1fed4 100644
--- a/ts/private/ts_lib.bzl
+++ b/ts/private/ts_lib.bzl
@@ -70,6 +70,18 @@ https://docs.aspect.build/rulesets/aspect_rules_js/docs/js_library#deps for more
         default = -1,
         values = [-1, 0, 1],
     ),
+    "declaration_transpile": attr.int(
+        doc = """\
+        Whether tsc should be used to produce .d.ts outputs
+
+        Values are:
+        - -1: Error if --@aspect_rules_ts//ts:default_to_tsc_transpiler not set, otherwise transpile
+        - 0: Do not transpile
+        - 1: Transpile
+        """,
+        default = -1,
+        values = [-1, 0, 1],
+    ),
     "tsc": attr.label(
         doc = "TypeScript compiler binary",
         mandatory = True,
diff --git a/ts/private/ts_project.bzl b/ts/private/ts_project.bzl
index 0d73be7..c95c3ec 100644
--- a/ts/private/ts_project.bzl
+++ b/ts/private/ts_project.bzl
@@ -64,8 +64,8 @@ def _ts_project_impl(ctx):
     typings_out_dir = ctx.attr.declaration_dir or ctx.attr.out_dir
     js_outs = _lib.declare_outputs(ctx, [] if ctx.attr.transpile == 0 else _lib.calculate_js_outs(srcs, ctx.attr.out_dir, ctx.attr.root_dir, ctx.attr.allow_js, ctx.attr.resolve_json_module, ctx.attr.preserve_jsx, ctx.attr.no_emit, ctx.attr.emit_declaration_only))
     map_outs = _lib.declare_outputs(ctx, [] if ctx.attr.transpile == 0 else _lib.calculate_map_outs(srcs, ctx.attr.out_dir, ctx.attr.root_dir, ctx.attr.source_map, ctx.attr.preserve_jsx, ctx.attr.no_emit, ctx.attr.emit_declaration_only))
-    typings_outs = _lib.declare_outputs(ctx, _lib.calculate_typings_outs(srcs, typings_out_dir, ctx.attr.root_dir, ctx.attr.declaration, ctx.attr.composite, ctx.attr.allow_js, ctx.attr.no_emit))
-    typing_maps_outs = _lib.declare_outputs(ctx, _lib.calculate_typing_maps_outs(srcs, typings_out_dir, ctx.attr.root_dir, ctx.attr.declaration_map, ctx.attr.allow_js, ctx.attr.no_emit))
+    typings_outs = _lib.declare_outputs(ctx, [] if ctx.attr.declaration_transpile == 0 else _lib.calculate_typings_outs(srcs, typings_out_dir, ctx.attr.root_dir, ctx.attr.declaration, ctx.attr.composite, ctx.attr.allow_js, ctx.attr.no_emit))
+    typing_maps_outs = _lib.declare_outputs(ctx, [] if ctx.attr.declaration_transpile == 0 else _lib.calculate_typing_maps_outs(srcs, typings_out_dir, ctx.attr.root_dir, ctx.attr.declaration_map, ctx.attr.allow_js, ctx.attr.no_emit))
 
     validation_outs = []
     if ctx.attr.validate:
@@ -181,6 +181,10 @@ See https://github.com/aspect-build/rules_ts/issues/361 for more details.
     if len(outputs) > 0 and ctx.attr.transpile == -1 and not ctx.attr.emit_declaration_only and not options.default_to_tsc_transpiler:
         fail(transpiler_selection_required)
 
+    # TODO: Determine if this is necessary
+    # if len(outputs) > 0 and ctx.attr.declaration_transpile == -1 and ctx.attr.emit_declaration_only and not options.default_to_tsc_transpiler:
+    #     fail(declaration_transpiler_selection_required)
+
     output_types = typings_outs + typing_maps_outs + typings_srcs
     use_tsc_noemit = ctx.attr.no_emit or (ctx.attr.transpile == 0 and not ctx.attr.declaration)
```

Reply:
That looks about right to me. I think we can just use a boolean though, or even just set `no_emit` from the macro.

I don't think we want the `fail` though, at least for now. That is more important for .js and I don't think we have a recommended dts transpiler at this time like we do with the js transpiler.